src/Microsoft.VisualStudio.Web.CodeGeneration.MsBuild/RoslynWorkspace.cs

```diff
@@ -0,0 +1,118 @@
+using System.IO;
+using Microsoft.CodeAnalysis.Text;
+using System.Text;
+using System.Reflection.PortableExecutable;
+using Microsoft.Extensions.DependencyModel;
+
+namespace Microsoft.VisualStudio.Web.CodeGeneration.MsBuild
+{
+    public class RoslynWorkspace : Workspace
+    {
+        private Dictionary<string, AssemblyMetadata> _cache = new Dictionary<string, AssemblyMetadata>();
+        private HashSet<string> _projectReferences = new HashSet<string>();
+        //MsBuildProjectContext _context;
+        public RoslynWorkspace(MsBuildProjectContext context, string configuration = "debug") : base(MefHostServices.DefaultHost, "Custom")
+        {
+            var id = AddProject(context, configuration);
+            AddMetadataReferences(context, id);
+        }
+
+        private void AddMetadataReferences(MsBuildProjectContext context, ProjectId id)
+        {
+            var libraryExporter = context.CreateLibraryExporter();
+            var exports = libraryExporter.GetExports() ?? new List<Library>();
+            
+            foreach(var export in exports)
+            {
+                if(!_projectReferences.Contains(export.Name))
+                {
+                    var compLibrary = export as CompilationLibrary;
+                    if(compLibrary != null)
+                    {
+                        // TODO How to get resolved path ?
+                        OnMetadataReferenceAdded(id, GetMetadataReference(""));
+                    }
+                }
+            }
+        }
+
+        private ProjectId AddProject(MsBuildProjectContext context, string configuration)
+        {
+            var projectInfo = ProjectInfo.Create(
+                ProjectId.CreateNewId(),
+                VersionStamp.Create(),
+                context.ProjectName,
+                context.ProjectName,
+                LanguageNames.CSharp,
+                context.ProjectFullPath);
+
+            _projectReferences.Add(context.ProjectName);
+
+            OnProjectAdded(projectInfo);
+
+            foreach (var file in context.ProjectFile.SourceFiles)
+            {
+                var filePath = Path.IsPathRooted(file.EvaluatedInclude)
+                    ? file.EvaluatedInclude :
+                    Path.Combine(Path.GetDirectoryName(context.ProjectFullPath), file.EvaluatedInclude);
+                AddSourceFile(projectInfo, filePath);
+            }
+
+            //TODO: Look at project dependencies, and add metadata references
+            // Also need to figure out how to add the shared sources.
+
+            foreach(var project in context.ProjectFile.ProjectReferences)
+            {
+                var projPath = project.EvaluatedInclude;
+                if (!Path.IsPathRooted(projPath))
+                {
+                    Path.Combine(Path.GetDirectoryName(context.ProjectFullPath), projPath);
+                }
+                var id = AddProject(new MsBuildProjectContext(projPath, configuration), configuration);
```

Comment:
You should add a check to avoid re-adding projects that are already in the workspace. (I had to do this in https://github.com/natemcmaster/AspNetSourceBrowser/blob/f00de862cc4ad36d49de3a13c9119e7cfa247123/src/Common/ProjectJsonWorkspace.cs#L63-L67). Otherwise, this can eat up memory by re-adding exiting projects. (FWIW Microsoft.DotNet.ProjectModel.Workspaces also has this bug...)